cassiopeia/simulator/Cas9LineageTracingDataSimulator.py

```diff
@@ -0,0 +1,202 @@
+
+            if tree.is_root(node):
+                tree.set_character_states = [0]*number_of_characters
+                continue
+
+            parent = tree.get_parent(node) 
+            t = tree.get_time(node) - tree.get_time(parent)
+
+            p = 1 - (np.exp(-t * self.mutation_rate))
+    
+            character_array = tree.get_character_states(parent)
+            open_sites = [c for c in range(len(character_array)) if character_array[c] == 0]
+
+            number_of_mutations = np.random.binomial(len(open_sites), p)
+            new_cuts = np.random.choice(open_sites, number_of_mutations)
+            
+            # collapse cuts that are on the same cassette
+            character_array, cuts_remaining = self.collapse_sites(character_array, new_cuts)
+            
+            # introduce new states at cut sites
+            character_array = self.introduce_states(character_array, cuts_remaining)
+            
+            # silence cassettes
+            character_array = self.silence_cassettes(character_array)
+
+            tree.set_character_states(node, character_array)
+
+
+    def collapse_sites(self, character_array: List[int], cuts: List[int]) -> Tuple[List[int], List[int]]:
+        """Collapses cassettes.
+
+        Given a character array and a new set of cuts that Cas9 is inducing,
+        this function will infer which cuts occur within a given cassette and
+        collapse the sites between the two cuts.
+
+        Args:
+            character_array: Character array in progress
+            cuts: Sites in the character array that are being cut.
+
+        Returns:
+            The updated character array and sites that are not part of a
+                cassette collapse.
+        """
+
+        cassettes = self.get_cassettes()
+        cut_to_cassette = np.digitize(cuts, cassettes)
+        cuts_remaining =[]
+        for cassette in np.unique(cut_to_cassette):
+            cut_indices = np.where(cut_to_cassette == cassette)[0]
+            if len(cut_indices) > 1:
+                cuts = new_cuts[cut_indices]
+                left, right = np.min(cuts), np.max(cuts)
+                for site in range(left, right+1):
+                    character_array[site] = -1
+            else:
+                cuts_remaining.append(cut_indices[0])
+        
+        return character_array, cuts_remaining
+
+    def introduce_states(self, character_array: List[int], cuts: List[int]) -> List[int]:
+        """Adds states to character array.
+
+        New states are added to the character array at the predefined cut
+        locations.
+
+        Args:
+            character_array: Character array
+            cuts: Loci being cut
+
+        Returns:
+            An updated character array.
+        """
+
+        for i in cuts:
+            state = self.random.choice(list(self.mutation_priors.keys()), 1, p=list(self.mutation_priors.values()))
```

Comment:
This line is hard to understand, can we split it into multiple lines, each of which is easy to understand?

Reply:
I fixed this in a recent commit, how does it look now?